### ISSUE TITLE
Would be nice to have more info about the project and it's goals

### ISSUE THREAD

**Issue by ryanford-frontend** [hishamhm/tl#24]
Is this project related to typed lua, titan lang or pallene? Is it meant to be a more minimal version? Like Lua + typing vs titan/pallene opting to be "companion languages"?

I'm very interested in the project based on the talk, but I would like to see more about the intended features and design goals.

**Comment by hishamhm** [comment]
Good question, things are understandably confusing at this point!

## Other related projects

* [Typed Lua](https://github.com/andremm/typedlua) was a research project started as @andremm's PhD thesis at PUC-Rio, co-advised by @mascarenhas which also got contributions from some students over time. It was a big exploration of optional typing for Lua and lots of lessons were learned there (see the various published papers!). The focus being on research results, the implementation itself for most of the time remained more of a proof-of-concept of this research, rather than a practical tool. The original developers moved on to other things, and the repo hasn't had updates in a year.
* [Titan](https://github.com/titan-lang/titan) started as a community project between myself (maintainer of [LuaRocks](https://luarocks.org) and a PUC-Rio alumnus), the Typed Lua devs mentioned above, plus grad students at PUC-Rio @hugomg and @gligneul who at the time were studying optional typing and compilation techniques. The idea was to join forces and build something that would make everyone happy: I wanted a Lua for programming-in-the-large, the Typed Lua people wanted a typed dialect of Lua, and grad students wanted a high-performance Lua. The "companion language" concept seemed to fit the bill, being a "Lua-ish language you'd use instead of falling back to C or doing contortions to please the JIT compiler gods".
* [Pallene](https://github.com/pallene-lang/pallene) started as a fork of Titan because, as you can guess, design-by-committee is challenging, especially when people in the committee have different goals. :) Since grad students work under a deadline to produce results, the ideal situation was to have the academic side of Titan fork into its own project, which is currently active, healthy and continues with the evolution of the "companion language" concept. Titan went dormant as the others moved on: and as far as I'm concerned, Titan is now dead but I hope it will prove to have been a worthwhile kickstarting effort once Pallene starts to bear fruit.

## This project's goals

In the meantime, my day job involves developing [two large](http://konghq.com) [Lua applications](http://luarocks.org), so yes, I _still_ want a practical Lua dialect for programming-in-the-large: so tl was born!

The goal of tl is to be a dialect of Lua as much as TypeScript is a dialect of JavaScript. So yes, it is on the one hand a different language (there are new keywords such as `global` for example) but on the other hand it is pretty much recognizable, as you said, as "Lua + typing".

tl aims to integrate to pretty much any Lua environment, since it's a "transpiler" that generates plain Lua and has no dependencies. I hope to support at least both the latest PUC-Rio Lua and the latest LuaJIT as output targets.

Minimalism (for some vague definition of minimalism which I plan to clarify in a future talk soon!) is a design goal for both conceptual and practical reasons: conceptually to match the nature of Lua, and practical so that I can manage developing it. :)

My very first concrete goal for tl's development was to have the compiler typecheck itself; that was achieved already: tl is written in tl.

My next big goal, which is I hope to achieve sometime next year, is to have it typecheck the LuaRocks source code. That's something I wanted since the Typed Lua days back in 2015. That's a big goal and once I get there I'll dare call this "production ready", since it's used in a real-world program, though it should be usable even before we get there!

## In summary

So **tl;dr:**

Typed Lua and Titan aren't active, Pallene is the companion language project, and tl is a transpiler for Lua plus types.

tl (we really need a better name!) is a project I'm writing in hopes it will be useful for myself and hopefully others. I'm trying to keep it small and long-term manageable, and would love to see a community of users and contributors grow around it over time!
